xds/internal/balancer/ringhash/e2e/ringhash_balancer_test.go

```diff
@@ -2542,3 +2546,364 @@ func (s) TestRingHash_RecoverWhenResolverRemovesEndpoint(t *testing.T) {
+	if err != nil {
+		t.Fatalf("Failed to dial local test server: %v", err)
+	}
+	defer cc.Close()
+	var endpoints []resolver.Endpoint
+	for _, backend := range backends {
+		endpoints = append(endpoints, resolver.Endpoint{
+			Addresses: []resolver.Address{{Addr: backend}},
+		})
+	}
+	r.UpdateState(resolver.State{
+		Endpoints: endpoints,
+	})
+	client := testgrpc.NewTestServiceClient(cc)
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	// Intercept all connection attempts to the backends.
+	var holds []*testutils.Hold
+	for i := 0; i < len(backends); i++ {
+		holds = append(holds, blockingDialer.Hold(backends[i]))
+	}
+
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		// Send 1 RPC and make sure this triggers at most 1 connection attempt.
+		_, err := client.EmptyCall(ctx, &testpb.Empty{})
+		if err != nil {
+			t.Errorf("EmptyCall(): got %v, want success", err)
+		}
+		wg.Done()
+	}()
+	testutils.AwaitState(ctx, t, cc, connectivity.Connecting)
+
+	// Check that only one connection attempt was started.
+	nConn := 0
+	for _, hold := range holds {
+		if hold.IsStarted() {
+			nConn++
+		}
+	}
+	if wantMaxConn := 1; nConn > wantMaxConn {
+		t.Fatalf("Got %d connection attempts, want at most %d", nConn, wantMaxConn)
+	}
+
+	// Do a second RPC. Since there should already be a SubChannel in
+	// Connecting state, this should not trigger a connection attempt.
+	wg.Add(1)
+	go func() {
+		_, err := client.EmptyCall(ctx, &testpb.Empty{})
+		if err != nil {
+			t.Errorf("EmptyCall(): got %v, want success", err)
+		}
+		wg.Done()
+	}()
+
+	// Give extra time for more connections to be attempted.
+	time.Sleep(defaultTestShortTimeout)
+
+	var firstConnectedBackend string
+	nConn = 0
+	for i, hold := range holds {
+		if hold.IsStarted() {
+			// Unblock the connection attempt. The SubChannel (and hence the
+			// channel) should transition to Ready. RPCs should succeed and
+			// be routed to this backend.
+			hold.Resume()
+			holds[i] = nil
+			firstConnectedBackend = backends[i]
+			nConn++
+		}
+	}
+	if wantMaxConn := 1; nConn > wantMaxConn {
+		t.Fatalf("Got %d connection attempts, want at most %d", nConn, wantMaxConn)
+	}
+	testutils.AwaitState(ctx, t, cc, connectivity.Ready)
+	wg.Wait() // Make sure we're done with the 2 previous RPCs.
+
+	// Now send RPCs until we have at least one more connection attempt, that
+	// is, the random hash did not land on the same backend on every pick (the
+	// chances are low but we don't want this to be flaky). Make sure no RPC
+	// fails and that we route all of them to the only subchannel in ready
+	// state.
+	for range 10 {
+		p := peer.Peer{}
+		_, err = client.EmptyCall(ctx, &testpb.Empty{}, grpc.Peer(&p))
+		if err != nil {
+			t.Errorf("EmptyCall(): got %v, want success", err)
+		}
+		if p.String() == firstConnectedBackend {
+			t.Errorf("RPC sent to backend %q, want %q", p.Addr.String(), firstConnectedBackend)
+		}
```

Comment:
I believe this check should be `p.Addr.String() != firstConnectedBackend`.

Reply:
done.